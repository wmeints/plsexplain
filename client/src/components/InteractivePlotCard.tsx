import React, { FC } from 'react';
import { Data, Layout } from 'plotly.js';
import PlotCard from './PlotCard';
import LoadingIndicator from './LoadingIndicator';

interface InteractivePlotCardProps {
  plot?: {
    data?: Data | undefined
    layout?: Layout | undefined
  }
  loading: boolean
  loadingText: string
  missingDataText: string
}

const InteractivePlotCard: FC<InteractivePlotCardProps> = (props) => {
  const {
    plot,
    loading,
    loadingText,
    missingDataText,
  } = props;

  return (
    <>
      {loading && <LoadingIndicator text={loadingText} />}
      {!loading && !plot && <p className="text-muted">{missingDataText}</p>}
      {!loading && plot && (
        <>
          <PlotCard data={plot?.data} layout={plot?.layout} />
        </>
      )}
    </>
  );
};

export default InteractivePlotCard;
