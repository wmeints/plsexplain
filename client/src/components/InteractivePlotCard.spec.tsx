/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { render } from 'enzyme';
import InteractivePlotCard from './InteractivePlotCard';
import plotData from '../__samples__/feature-importance.json';

test('renders loading state correctly', () => {
  const wrapper = render(
    <InteractivePlotCard
      loading
      loadingText="Loading data"
      missingDataText=""
      plot={plotData as any}
    />,
  );

  expect(wrapper).toMatchSnapshot();
});

test('renders normal state correctly', () => {
  const wrapper = render(
    <InteractivePlotCard
      loading={false}
      loadingText="Loading data"
      missingDataText=""
      plot={plotData as any}
    />,
  );

  expect(wrapper).toMatchSnapshot();
});

test('renders normal state correctly', () => {
  const wrapper = render(
    <InteractivePlotCard
      loading={false}
      loadingText="Loading data"
      missingDataText="This is missing text"
      plot={undefined}
    />,
  );

  expect(wrapper).toMatchSnapshot();
});
