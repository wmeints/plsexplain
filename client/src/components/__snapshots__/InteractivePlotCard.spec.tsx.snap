// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders loading state correctly 1`] = `
LoadedCheerio {
  "0": Node {
    "attribs": Object {
      "aria-hidden": "true",
      "class": "spinner-border spinner-border-sm mr-2",
      "role": "status",
    },
    "children": Array [],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": Node {
      "attribs": Object {
        "data-test": "loadingText",
      },
      "children": Array [
        Node {
          "data": "Loading data",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "text",
        },
      ],
      "name": "span",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": Node {
        "children": Array [
          [Circular],
          [Circular],
        ],
        "name": "root",
        "next": null,
        "parent": null,
        "prev": null,
        "type": "root",
      },
      "prev": [Circular],
      "type": "tag",
      "x-attribsNamespace": Object {
        "data-test": undefined,
      },
      "x-attribsPrefix": Object {
        "data-test": undefined,
      },
    },
    "parent": Node {
      "children": Array [
        [Circular],
        Node {
          "attribs": Object {
            "data-test": "loadingText",
          },
          "children": Array [
            Node {
              "data": "Loading data",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "data-test": undefined,
          },
          "x-attribsPrefix": Object {
            "data-test": undefined,
          },
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "aria-hidden": undefined,
      "class": undefined,
      "role": undefined,
    },
    "x-attribsPrefix": Object {
      "aria-hidden": undefined,
      "class": undefined,
      "role": undefined,
    },
  },
  "1": Node {
    "attribs": Object {
      "data-test": "loadingText",
    },
    "children": Array [
      Node {
        "data": "Loading data",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "span",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        Node {
          "attribs": Object {
            "aria-hidden": "true",
            "class": "spinner-border spinner-border-sm mr-2",
            "role": "status",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "aria-hidden": undefined,
            "class": undefined,
            "role": undefined,
          },
          "x-attribsPrefix": Object {
            "aria-hidden": undefined,
            "class": undefined,
            "role": undefined,
          },
        },
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": Node {
      "attribs": Object {
        "aria-hidden": "true",
        "class": "spinner-border spinner-border-sm mr-2",
        "role": "status",
      },
      "children": Array [],
      "name": "div",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": [Circular],
      "parent": Node {
        "children": Array [
          [Circular],
          [Circular],
        ],
        "name": "root",
        "next": null,
        "parent": null,
        "prev": null,
        "type": "root",
      },
      "prev": null,
      "type": "tag",
      "x-attribsNamespace": Object {
        "aria-hidden": undefined,
        "class": undefined,
        "role": undefined,
      },
      "x-attribsPrefix": Object {
        "aria-hidden": undefined,
        "class": undefined,
        "role": undefined,
      },
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "data-test": undefined,
    },
    "x-attribsPrefix": Object {
      "data-test": undefined,
    },
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 2,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`renders normal state correctly 1`] = `
LoadedCheerio {
  "0": Node {
    "attribs": Object {
      "class": "card",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "class": "card-body",
        },
        "children": Array [
          Node {
            "attribs": Object {
              "style": "position:relative;display:inline-block",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "style": undefined,
            },
            "x-attribsPrefix": Object {
              "style": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`renders normal state correctly 2`] = `
LoadedCheerio {
  "0": Node {
    "attribs": Object {
      "class": "text-muted",
    },
    "children": Array [
      Node {
        "data": "This is missing text",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "p",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
